### PR TITLE
CW decoder

### PR DESCRIPTION
**Pull request by CanoRoger**

### CODE REVIEW

**Comment by CanoRoger** [comment]
starcoder_cw_decoder

File: gr-starcoder/lib/cw_to_symbol_impl.cc [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+
+      if(!d_const_val || !d_tmp || !d_out) {
+        throw std::runtime_error("cw_to_symbol: Could not allocate memory");
+      }
+
+      for(i = 0; i < d_window_size; i++) {
+        d_const_val[i] = threshold;
+      }
+      set_history(d_window_size);
+      // printf("Initial threshold value used = %f", threshold); //////////////////////
```
Comment (reiinakano) [inline gr-starcoder/lib/cw_to_symbol_impl.cc]: Same as above but you can use `GR_LOG_DEBUG` for this

File: gr-starcoder/lib/cw_to_symbol_impl.cc [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+      /* During idle state search for a possible trigger */
+      if(d_dec_state == NO_SYNC) {
+        for(i = 0; i < (size_t)noutput_items; i++) {
+          /*
+           * Clamp the input so the window mean is not affected by strong spikes
+           * Good luck understanding this black magic shit!
+           */
+          triggered = is_triggered(in_old + i, d_window_size);
+          if(triggered) {
+            //LOG_DEBUG("Triggered!");
```
Comment (reiinakano) [inline gr-starcoder/lib/cw_to_symbol_impl.cc]: delete

File: gr-starcoder/lib/cw_to_symbol_impl.cc [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+      /* From now one, we handle the input in multiples of a window */
+      for (i = 0; i < (size_t)noutput_items / d_window_size; i++) {
+        triggered = is_triggered(in + i * d_window_size, d_window_size);
+        switch(d_dec_state) {
+          case SEARCH_DOT:
+            if(triggered) {
+              d_window_cnt++;
+              if(d_window_cnt > d_dot_windows_num) {
+                set_long_on();
+                //LOG_DEBUG("Going to search for long sequence");
```
Comment (reiinakano) [inline gr-starcoder/lib/cw_to_symbol_impl.cc]: delete

File: gr-starcoder/lib/cw_to_symbol_impl.cc [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+              }
+            }
+            else {
+              if(check_conf_level(d_window_cnt, d_dot_windows_num)) {
+                //LOG_DEBUG("DOT");
```
Comment (reiinakano) [inline gr-starcoder/lib/cw_to_symbol_impl.cc]: delete

File: gr-starcoder/lib/cw_to_symbol_impl.cc [hunk 5]
```diff
@@ -0,0 +1,387 @@
...
+                send_symbol_msg(MORSE_DOT);
+              }
+              //LOG_DEBUG("Going to search for space: win cnt %lu", d_window_cnt);
```
Comment (reiinakano) [inline gr-starcoder/lib/cw_to_symbol_impl.cc]: delete

File: gr-starcoder/lib/cw_to_symbol_impl.cc [hunk 6]
```diff
@@ -0,0 +1,387 @@
...
+              set_search_space ();
+            }
+            break;
+          case SEARCH_DASH:
+            if(triggered) {
+              d_window_cnt++;
+            }
+            else{
+              if(check_conf_level(d_window_cnt, d_dash_windows_num)) {
+                //LOG_DEBUG("DASH");
```
Comment (reiinakano) [inline gr-starcoder/lib/cw_to_symbol_impl.cc]: delete all commented out logs in this method

File: gr-starcoder/lib/morse_decoder_impl.cc [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <gnuradio/io_signature.h>
+#include "morse_decoder_impl.h"
+//#include <satnogs/log.h>
```
Comment (reiinakano) [inline gr-starcoder/lib/morse_decoder_impl.cc]: delete

File: gr-starcoder/lib/morse_decoder_impl.cc [hunk 8]
```diff
@@ -0,0 +1,141 @@
...
+            message_port_pub (pmt::mp ("out"),
+                              pmt::make_blob (str.c_str (), str.length ()));
+          }
+          break;
+        case MORSE_INTRA_SPACE:
+          /*Ignore it */
+          break;
+        default:
+          // LOG_ERROR("Unknown Morse symbol");
+	  printf("Unknown Morse symbol");
```
Comment (reiinakano) [inline gr-starcoder/lib/morse_decoder_impl.cc]: GR_LOG_ERROR then remove the printf.

File: gr-starcoder/lib/morse_tree.cc [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include <gnuradio/io_signature.h>
+#include <starcoder/morse_tree.h>
+// #include <satnogs/log.h>
```
Comment (reiinakano) [inline gr-starcoder/lib/morse_tree.cc]: delete

File: flowgraphs/starcoder_cw_decoder.grc [hunk 10]
```diff
@@ -0,0 +1,1901 @@
...
+      <key>_coordinate</key>
+      <value>(848, 748)</value>
+    </param>
+    <param>
+      <key>_rotation</key>
+      <value>0</value>
+    </param>
+    <param>
+      <key>id</key>
+      <value>cw_wpm</value>
```
Comment (reiinakano) [inline flowgraphs/starcoder_cw_decoder.grc]: Can you change the parameter name  to `wpm`

File: flowgraphs/starcoder_cw_decoder.grc [hunk 11]
```diff
@@ -0,0 +1,1901 @@
+<?xml version='1.0' encoding='utf-8'?>
```
Comment (reiinakano) [inline flowgraphs/starcoder_cw_decoder.grc]: Rename this whole file to just cw_decoder.grc. You can delete the previous cw_decoder.grc

**Pull request merged**